test/java/seedu/addressbook/logic/LogicTest.java

```diff
@@ -87,10 +88,13 @@ private void assertCommandBehavior(String inputCommand,
             assertEquals(lastShownList, r.getRelevantPersons().get());
         }
 
-        //Confirm the state of data is as expected
-        assertEquals(expectedAddressBook, addressBook);
-        assertEquals(lastShownList, logic.getLastShownList());
-        assertEquals(addressBook, saveFile.load());
+        //Confirm the state of data is as expected, if command is capable of mutation
+        Command command = new Parser().parseCommand(inputCommand);
+        if(command.isMutating()) {
```

Comment:
How about the else branch? i.e. when the command is not mutating? What is the expected result to be asserted?

Reply:
Hey, sorry I'm having some trouble understanding.

If the command is not mutating, I figure we would only need to check the addressbook against the storagefile to ensure that no changes have been made, which I've done in the latest commit. This is somehow causing a few tests to fail.

Reply:
@philemontan The checking is a bit non-obvious. To elaborate a bit more, the isMutating() method is implemented to ensure that the saving of the storage file only happens when a mutating command is executed. Therefore, checking the contents of the storage file will not be sufficient as the storage file can be saved repeatedly without making changes. What is desired is however to check the storage file is untouched with a non-mutating command. This has more to do with the timestamp of the file object, i.e. checking the last modified attribute or whether it is still the same file object after all. You can refer to this PR [for](https://github.com/nus-cs2103-AY1718S1/addressbook-level3/pull/216) more details and also correcting your test failures. His way of handling is quite elegant.

Reply:
What i gather from that PR, is that this: 

        } else {
            assertNotSame(expectedAddressBook, saveFile.load());
        }

is supposed to check that the storage file is untouched.

So if the command is non-mutating, then the expectedAddressBook should be different from the current saved storage file?

Does this mean non-mutating commands pass a reference to an altered version of the addressbook into the assertCommandBehaviour method? This seems rather unintuitive to me, since i was expecting the actual, expected and storagefile to all be the same.

Reply:
@philemontan Oops, I sent the wrong link. Sorry for this. The PR which I was referring to compares saveFile.load() with an empty Addressbook object. It does some meaningful checking as we start with an empty Addressbook object for each test. However, it is still not logically sound to achieve what we want. Using Java File lastModified() will be more suitable in this case. For your errors, you should not use assertEquals with saveFile.load() and addressBook for the not-mutating commands. Read the failed tests file carefully. You will see that addressBook is modified in these tests(it is not empty), but no saving is ever done, i.e. saveFile is not updated accordingly. Since we start with empty saveFile for every test, the assertion will always fail when you compare non-empty with empty objects. I also notice your return value of isMutating() for the ClearCommand class is wrong. Change it to true instead. Sorry for the earlier confusions. This additional checking is actually considered as some bonus for strong students, as it is good to go through some thinking process :) If you have any other questions, feel free to contact me here or via Slack!

Reply:
@OuyangDanwen Hey, thanks for being patient and explaining! I've managed to implement this in the else branch already. I'll be closing this PR. Thanks again, learnt quite a bit!